### PR TITLE
ERC20 Permit Component

### PR DESCRIPTION
**Pull request by immrsd**
Implements [ERC20Permit extension](https://github.com/OpenZeppelin/cairo-contracts/issues/1103)

#### PR Checklist

- [x] Tests
- [ ] Documentation in a separate PR
- [x] Added entry to CHANGELOG.md
- [ ] Tried the feature on a public network

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+
+- ERC20Permit component and preset
```
Comment (ericnordelo) [inline CHANGELOG.md:14]:
```suggestion
- ERC20Permit component and preset (#1140)
```

File: packages/presets/src/erc20_permit.cairo [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+/// transactions.
+///
+/// This implementation features a fixed-supply model and the initial owner is specified in the
+/// constructor.
+///
+/// The preset implements SNIP12Metadata with hardcoded values. These values are part of a
+/// signature (following SNIP-12 standard) used for ERC20Permit functionality. It's crucial that the
+/// SNIP12Metadata name remains unique to avoid confusion and potential security issues.
+///
+/// For more complex or custom contracts, use Wizard for Cairo
```
Comment (ericnordelo) [inline packages/presets/src/erc20_permit.cairo:18]: Will we have this release together with a Wizard update?

**Review by ericnordelo — COMMENTED** [review]
Hey Alejandro, the implementation is looking good, but I think we should refactor this into an extra embeddable implementation in the ERC20Component, instead of an extra component, since this would reduce the boilerplate in the final contract, and a component without storage members and events can be just an extra module.

**Review by andrew-fleming — COMMENTED** [review]
Looking good, sir! I do agree with Eric with having ERC20Permit as an embeddable impl inside of ERC20Component. This should be a little easier to integrate into a contract. Aside from this, I left a few comments :)

File: packages/presets/src/erc20_permit.cairo (hunk 2 above)
Comment (immrsd) [reply packages/presets/src/erc20_permit.cairo]: Removed ERC20Permit preset and this message also as a result

File: packages/token/src/tests/erc20/test_erc20_permit.cairo [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    assert_valid_nonce(owner, nonce + 1);
+}
+
+#[test]
+fn test_valid_permit_other_data() {
+    let mut data = TEST_DATA();
+    data.spender = constants::OTHER();
+    data.amount = constants::TOKEN_VALUE_2;
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mock = setup(data);
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/tests/erc20/test_erc20_permit.cairo:104]: Frankly, I don't think this would be helpful. There would be only 2 test cases covered by this helper, all other cases have slight differences

File: packages/token/src/tests/mocks/erc20_permit_mocks.cairo [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        ERC20Event: ERC20Component::Event,
+        #[flat]
+        ERC20PermitEvent: ERC20PermitComponent::Event,
+        #[flat]
+        NoncesEvent: NoncesComponent::Event
+    }
+
+    /// Required for hash computation.
+    pub(crate) impl SNIP12MetadataImpl of SNIP12Metadata {
+        fn name() -> felt252 {
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/tests/mocks/erc20_permit_mocks.cairo:44]: Yes, now I import the implementation to use the same `name` and `version` in the test setup function

File: packages/presets/src/erc20_permit.cairo [hunk 5]
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.16.0 (presets/erc20_permit.cairo)
+
+/// # ERC20Permit Preset
+///
+/// The ERC20Permit preset integrates standard ERC20 token functionality with the ERC20Permit
+/// extension, as defined by EIP-2612. This preset allows for token approvals via off-chain
+/// signatures, thus enhancing transaction efficiency by reducing the need for on-chain approval
+/// transactions.
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/presets/src/erc20_permit.cairo]: Decided to go without the preset. It doesn't make sense to have it, especially once it was refactored to be a trait instead of a component

File: packages/token/src/erc20/extensions/erc20_permit/erc20_permit.cairo [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+///   - The address of the owner
+///   - The parameters specified in the `approve` function (spender and amount)
+///   - The address of the token contract itself
+///   - A nonce, which must be unique for each operation, incrementing after each use to prevent
+///   reuse of the signature - The chain ID, which protects against cross-chain replay attacks
+///
+/// EIP-2612: https://eips.ethereum.org/EIPS/eip-2612
+/// SNIP-12:  https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md
+#[starknet::component]
+pub mod ERC20PermitComponent {
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/extensions/erc20_permit/erc20_permit.cairo]: Completely agree, refactored it to be a trait of the ERC20 component

**Review by andrew-fleming — COMMENTED** [review]
Nice job on the improvements! This looks almost ready to me. I left a small suggestion and a question

File: packages/token/src/erc20/extensions/erc20_permit.cairo
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/extensions/erc20_permit.cairo:1]: Actually, that was my initial decision while refactoring. But having two files under extensions `erc20_votes` and `erc20_permit_utils` seemed inconsistent and the "utils" suffix felt redundant. I don't have a strong opinion on this matter, I'm okay with both of the names

[the remark this answers is not in the document]
Comment (andrew-fleming) [reply packages/token/src/erc20/extensions/erc20_permit.cairo:1]: I agree, it's not great. I'm okay with the current `erc20_permit` for now. @ericnordelo any strong opinions on this?

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Added
+
+- ERC20Permit component and preset (#1140)
```
Comment (ericnordelo) [inline CHANGELOG.md:14]: We need to update this. Also adding subentries will be helpful to users, like the embeddable implementations added and the utilities.

File: packages/token/src/erc20/extensions/erc20_permit.cairo
[the remark this answers is not in the document]
Comment (ericnordelo) [reply packages/token/src/erc20/extensions/erc20_permit.cairo:1]: For consistency with the recent SNIP-9 addition, I would move the file under `erc20` but outside of `extensions`, and rename it to `snip12_utils.cairo`, since it only contains snip12 utilities. We should also move the `erc20_votes` snip12 utilities that are currently in the same file of the component, to that new file outside extensions.

We should probably have only one `snip12_utils` module per package, meaning that a refactor of SNIP-9 would also be required, but I think is worth doing that to improve/keep organization.

Something like:

```
token
└─── erc20
│  └─── extensions
│  │    │  erc20_votes.cairo
│  └─── snip12_utils
│  │    │  votes.cairo
│  │    │  permit.cairo
│  │  erc20.cairo
│  │  extensions.cairo
│  │  interface.cairo
│  │  snip12_utils.cairo
```

File: packages/token/src/erc20/interface.cairo [hunk 8]
```diff
@@ -43,6 +43,20 @@ pub trait IERC20CamelOnly<TState> {
...
 
+#[starknet::interface]
+pub trait IERC20Permit<TState> {
+    fn permit(
+        ref self: TState,
+        owner: ContractAddress,
+        spender: ContractAddress,
+        amount: u256,
+        deadline: u64,
+        signature: Array<felt252>
```
Comment (ericnordelo) [inline packages/token/src/erc20/interface.cairo:54]:
```suggestion
        signature: Span<felt252>
```

File: packages/token/src/erc20/interface.cairo [hunk 9]
```diff
@@ -107,3 +121,43 @@ pub trait ERC20VotesABI<TState> {
         ref self: TState, sender: ContractAddress, recipient: ContractAddress, amount: u256
     ) -> bool;
 }
+
+#[starknet::interface]
+pub trait ERC20PermitABI<TState> {
```
Comment (ericnordelo) [inline packages/token/src/erc20/interface.cairo:126]: We don't need to add an extra ABI trait, but include the new implementations in the ERC20ABI trait, even if the user can choose not to use some implementations, the ABI should contain the full ABI of the component.

File: packages/token/src/tests/erc20/test_erc20_permit.cairo [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+use openzeppelin_utils::cryptography::snip12::{StarknetDomain, StructHashStarknetDomainImpl};
+use openzeppelin_utils::serde::SerializedAppend;
+use snforge_std::signature::stark_curve::StarkCurveSignerImpl;
+use snforge_std::{
+    start_cheat_caller_address, start_cheat_block_timestamp, start_cheat_chain_id_global
+};
+use starknet::ContractAddress;
+
+//
+// Test Data
```
Comment (ericnordelo) [inline packages/token/src/tests/erc20/test_erc20_permit.cairo:19]:
```suggestion
// Constants
```

File: packages/token/src/tests/erc20/test_erc20_permit.cairo [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+//
+// Setup
+//
+
+fn setup(data: TestData) -> ERC20PermitABIDispatcher {
+    start_cheat_chain_id_global(data.chain_id);
+
+    utils::declare_and_deploy_at(
+        "DualCaseAccountMock", data.owner, array![data.key_pair.public_key]
+    );
```
Comment (ericnordelo) [inline packages/token/src/tests/erc20/test_erc20_permit.cairo:66]: Is something preventing us from using `COMPONENT_STATE` instead of deploying and dispatchers? Note that using COMPONENT_STATE should speed up running the tests.

**Review by ericnordelo — COMMENTED** [review]
Great job with the refactor @immrsd! Left some comments.

File: packages/token/src/erc20/erc20.cairo [hunk 12]
```diff
@@ -288,6 +296,104 @@ pub mod ERC20Component {
...
+        /// - `signature` must use the current nonce of the `owner`.
+        ///
+        /// Emits an `Approval` event.
+        fn permit(
+            ref self: ComponentState<TContractState>,
+            owner: ContractAddress,
+            spender: ContractAddress,
+            amount: u256,
+            deadline: u64,
+            signature: Array<felt252>
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/erc20.cairo:343]: We should favor Span over Array, but in this case we need Array (to make the `is_valid_signature` call to the owner's account, which expects a signature as `Array<felt252>`)

File: packages/token/src/erc20/erc20.cairo [hunk 13]
```diff
@@ -288,6 +296,104 @@ pub mod ERC20Component {
...
+    ///   - The address of the token contract itself
+    ///   - A nonce, which must be unique for each operation, incrementing after each use to prevent
+    ///   reuse of the signature - The chain ID, which protects against cross-chain replay attacks
+    ///
+    /// EIP-2612: https://eips.ethereum.org/EIPS/eip-2612
+    /// SNIP-12:  https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md
+    #[embeddable_as(ERC20PermitImpl)]
+    impl ERC20Permit<
+        TContractState,
+        impl ERC20: HasComponent<TContractState>,
```
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/erc20.cairo:319]: It's being used in the 5th step (approve):

```
let mut erc20_component = get_dep_component_mut!(ref self, ERC20);
erc20_component._approve(owner, spender, amount);
```

File: packages/token/src/erc20/interface.cairo (hunk 8 above)
Comment (immrsd) [reply packages/token/src/erc20/interface.cairo:54]: We have to require a signature in the format of an array

File: packages/token/src/tests/erc20/test_erc20_permit.cairo (hunk 11 above)
Comment (immrsd) [reply packages/token/src/tests/erc20/test_erc20_permit.cairo:66]: There's nothing preventing us from following the approach with `CONTRACT_STATE`, but there's also no strong reason why this approach is better. I reckon that having a contract deployed and interacting with it through dispatcher feels closer to the real-world usage scenario and also prevents from using any internal functions that 
are not supposed and even can't be called externally

Either way tests run pretty fast. That's compilation that takes most time

File: packages/token/src/erc20/extensions/erc20_permit.cairo
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/extensions/erc20_permit.cairo:1]: Restructured the files accordingly

File: packages/token/src/erc20/erc20.cairo (hunk 12 above)
[the remark this answers is not in the document]
Comment (ericnordelo) [reply packages/token/src/erc20/erc20.cairo:343]: Similar to the implementation in SRC9, we should receive the Span, and call into when calling `is_valid_signature`, since we only keep that Array for BC with the interface Id, but we should not continue adding more Array<felt252> when not needed. It would be nice if you could profile the calls using Span and into vs and Array to compare steps.

File: packages/token/src/erc20/interface.cairo (hunk 8 above)
Comment (ericnordelo) [reply packages/token/src/erc20/interface.cairo:54]: See https://github.com/OpenZeppelin/cairo-contracts/pull/1140#discussion_r1790077515

File: packages/token/src/erc20/interface.cairo (hunk 9 above)
Comment (ericnordelo) [reply packages/token/src/erc20/interface.cairo:126]: ERC20Mixin (and other mixins) implement the ABI of the component by coincidence, but not by definition. Mixin and ABIs are different things, even when they sometimes match.

A Mixin is a set of implementations that are commonly used together (as mentioned [here](https://docs.openzeppelin.com/contracts-cairo/0.17.0/components#mixins)) whose objective is to avoid verbosity. A component ABI trait, on the other hand, is intended as a single dispatcher for the component full interface, even if it doesn't have to be implemented all-together, as explained in this recently added note:

Screenshot 2024-10-07 at 14 04 24

We didn't follow this principle in Ownable and OwnableTwoStep because the same function (transfer ownership) with the same interface has different implementations that can be problematic if they are used thinking the other one is in place, so we favoured clarity over the rule. In this case, the permit implementation doesn't have a problematic intersection and can be added to the ABI as intended.

In this case, we can use `#[generate_trait]` for the mixin instead of implementing from the ABI trait, and we should add permit to the existing ABI.

Note that adding a different ABI per feature would create a lot of repetition in the interface file if the component grows bigger in features.

TLDR: Mixin shouldn't be necessarily pegged to component ABIs, since they are not defined as the same thing.

cc @andrew-fleming that may correct me if I recall something the wrong way from when we work on this definitions.

File: packages/token/src/tests/erc20/test_erc20_permit.cairo (hunk 11 above)
Comment (ericnordelo) [reply packages/token/src/tests/erc20/test_erc20_permit.cairo:66]: I had the same thoughts when we started following this logic, we already had a discussion about it and decided to test components as separate modules, since they are not deployed contracts, but modules to extend those. Declaring and deploying in tests take time and adds up while we continue adding more tests. I'm not against reopening the discussion about it, but for now, since we should always favor consistency, let's refactor into using COMPONENT_STATE as we've been doing when possible for component tests.

File: packages/token/src/erc20/interface.cairo (hunk 9 above)
Comment (andrew-fleming) [reply packages/token/src/erc20/interface.cairo:126]: This sounds accurate to me. Looking at the docs now, I do think we can improve our definition of ABI traits. We mention that they're useful for preset contracts; however, this isn't necessarily true anymore. We have a preset interfaces dir now and the ABI doesn't include the interface for upgrades. Further clarity and purpose would be helpful IMO

> TLDR: Mixin shouldn't be necessarily pegged to component ABIs, since they are not defined as the same thing.

Maybe it's worth using `#[generate_trait]` for all the mixins to establish a better differentiation

**Review by andrew-fleming — COMMENTED** [review]
Agreed with Eric's review, namely:
1. Span (+ into) > Array
2. ComponentState testing pattern > Dispatchers
    - If you feel strongly about favoring dispatchers for testing, let's discuss outside of the PR

File: packages/token/src/erc20/erc20.cairo (hunk 12 above)
[the remark this answers is not in the document]
Comment (immrsd) [reply packages/token/src/erc20/erc20.cairo]: Changed the input type to Span

File: CHANGELOG.md (hunk 7 above)
Comment (immrsd) [reply CHANGELOG.md]: Updated

File: packages/token/src/erc20/interface.cairo (hunk 9 above)
Comment (immrsd) [reply packages/token/src/erc20/interface.cairo]: `#[generate_trait]` wouldn't help here since the embeddable trait has to implement a starknet interface
I decoupled mixin impl from ABI and added a `IERC20Mixin` interface

File: packages/token/src/tests/erc20/test_erc20_permit.cairo (hunk 11 above)
Comment (immrsd) [reply packages/token/src/tests/erc20/test_erc20_permit.cairo:72]: Refactored to use component state instead of deploying a mock

File: packages/token/src/tests/erc20/test_erc20_permit.cairo [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+//
+// Expired signature
+//
+
+#[test]
+#[should_panic(expected: 'ERC20: expired permit signature')]
+fn test_invalid_sig_bad_deadline() {
+    let data = TEST_DATA();
+    let (owner, spender, amount, deadline) = (data.owner, data.spender, data.amount, data.deadline);
+    let mut state = setup(data);
```
Comment (ericnordelo) [inline packages/token/src/tests/erc20/test_erc20_permit.cairo:436]: This is an example of a test that doesn't use the `DualCaseAccountMock`, should we decouple the `setup_account` logic from the `setup` one?

File: packages/utils/src/cryptography/snip12.cairo [hunk 15]
```diff
@@ -32,7 +32,7 @@ pub trait OffchainMessageHash<T> {
     fn get_message_hash(self: @T, signer: ContractAddress) -> felt252;
 }
 
-impl StructHashStarknetDomainImpl of StructHash<StarknetDomain> {
```
Comment (ericnordelo) [inline packages/utils/src/cryptography/snip12.cairo:35]: Does this need to be pub? It should be enough to import the trait if needed, because the impl would be automatically in scope.

**Review by ericnordelo — APPROVED** [review]
Looking good @immrsd ! I left a small suggestion, but besides that I think it is ready to go.

@ggonzalez94 this will generate some conflicts with the Votes PR.

File: packages/token/src/tests/erc20/test_erc20_permit.cairo (hunk 14 above)
Comment (immrsd) [reply packages/token/src/tests/erc20/test_erc20_permit.cairo:436]: This particular test case doesn't require the account mock to be deployed. At the same time, I don't think we should decouple it, since this way we'll lose consistency over the test cases and we'll have to introduce an additional setup function that will be used in this only place
